File: backend-spring/src/main/java/org/example/cpt202music/model/entity/ReviewRecord.java
package org.example.cpt202music.model.entity;

import jakarta.persistence.*;
import lombok.Data;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
@Table(name = "review_records")
public class ReviewRecord {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "resource_id", nullable = false)
    private Resource resource;

    @ManyToOne
    @JoinColumn(name = "submission_id", nullable = false)
    private ResourceSubmission submission;

    @Column(name = "version_no", nullable = false)
    private Integer versionNo = 1;

    @ManyToOne
    @JoinColumn(name = "reviewer_id", nullable = false)
    private User reviewer;

    @Column(name = "action_description", nullable = false)
    @Enumerated(EnumType.STRING)
    private ActionDescription actionDescription;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private ReviewStatus status;

    @Column(name = "feedback_comment", columnDefinition = "TEXT")
    private String feedbackComment;

    @Column(name = "reviewed_at", nullable = false)
    private LocalDateTime reviewedAt;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @OneToMany(mappedBy = "reviewRecord", cascade = CascadeType.ALL)
    private Set<Feedback> feedbacks = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        reviewedAt = LocalDateTime.now();
        createdAt = LocalDateTime.now();
    }

    public enum ActionDescription {
        APPROVE, REJECT
    }

    public enum ReviewStatus {
        APPROVED, REJECTED
    }
}

File: backend/src/models/ReviewRecord.js
/**
 * ReviewRecord Model - #4 Module
 * Records each review history of a resource submission
 */

const { DataTypes } = require('sequelize')

const ACTION = {
  APPROVE: 'APPROVE',
  REJECT: 'REJECT'
}

const REVIEW_STATUS = {
  APPROVED: 'APPROVED',
  REJECTED: 'REJECTED'
}

module.exports = (sequelize) => {
  const ReviewRecord = sequelize.define('ReviewRecord', {
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true
    },
    resource_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      references: {
        model: 'resources',
        key: 'resource_id'
      }
    },
    submission_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      references: {
        model: 'resource_submission',
        key: 'submission_id'
      }
    },
    version_no: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 1
    },
    reviewer_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      references: {
        model: 'users',
        key: 'user_id'
      }
    },
    action_description: {
      type: DataTypes.ENUM(...Object.values(ACTION)),
      allowNull: false
    },
    status: {
      type: DataTypes.ENUM(...Object.values(REVIEW_STATUS)),
      allowNull: false
    },
    feedback_comment: {
      type: DataTypes.TEXT,
      allowNull: true,
      defaultValue: null
    },
    reviewed_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    },
    created_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    }
  }, {
    tableName: 'review_records',
    timestamps: false,
    indexes: [
      { fields: ['resource_id'] },
      { fields: ['reviewer_id'] }
    ]
  })

  ReviewRecord.associate = (models) => {
    ReviewRecord.belongsTo(models.Resource, {
      foreignKey: 'resource_id',
      as: 'resource'
    })

    ReviewRecord.belongsTo(models.ResourceSubmission, {
      foreignKey: 'submission_id',
      as: 'submission'
    })

    ReviewRecord.belongsTo(models.User, {
      foreignKey: 'reviewer_id',
      as: 'reviewer'
    })

    ReviewRecord.hasMany(models.Feedback, {
      foreignKey: 'review_record_id',
      as: 'feedbacks'
    })
  }

  ReviewRecord.ACTION = ACTION
  ReviewRecord.STATUS = REVIEW_STATUS

  return ReviewRecord
}
